Resources/doc/cookbook/recipe_hashid.rst

```diff
@@ -0,0 +1,118 @@
+
+Overriding the output of all URLs
+---------------------------------
+
+First you will need to encoding all the URLs that Sonata outputs,
+below I have chosen to use the ``encodeHex`` function but there is
+also an ``encode`` function. For more information on this see the
+documentation in the `HashId Symfony Bundle`_
+
+.. code:: php
+
+    class MyAdmin extends AbstractAdmin
+    {
+        /**
+         * @var HashId
+         */
+        private $hashId
+
+        /**
+         * @param string $code
+         * @param string $class
+         * @param string $baseControllerName
+         * @param HashId $hashId
+         */
+        public function __construct($code, $class, $baseControllerName, HashId $hashId)
+        {
+            parent::__construct($code, $class, $baseControllerName);
+            $this->hashId = $hashId
+        }
+
+        ... other configurations
+
+        /**
+         * @param string $name
+         * @param mixed $object
+         * @param array $parameters
+         * @param bool|int $absolute
+         * @return string
+         */
+        public function generateObjectUrl($name, $object, array $parameters = [], $absolute = UrlGeneratorInterface::ABSOLUTE_PATH)
+        {
+            $parameters['id'] = $this->hashId->encodeHex($this->getUrlsafeIdentifier($object));
+
+            return $this->generateUrl($name, $parameters, $absolute);
+        }
+    }
+
+
+Now your routes will change from ``/product/1/edit`` to
+``/product/af5/edit``, child routes will also be changed if you have
+chosen to use them e.g. ``product/af5/child/e6D/edit``. So now you will
+have to tell your controllers to decode these. For this implement a
+class that uses the ``RouteIdHandlerInterface`` to decode the id held
+in the request. For example
+
+.. code:: php
+
+    class HashIdRouteIdHandler implements RouteIdHandlerInterface
+    {
+        /**
+         * @var HashId
+         */
+        private $hashId
+
+        /**
+         * @param HashId $hashId
+         */
+        public function __construct(HashId $hashId)
+        {
+            $this->hashId = $hashId
+        }
+
+        /**
+         * @param Request $request
+         * @param AdminInterface $admin
+         *
+         * @return int|string
+         */
+        public function getIdFromRequest(Request $request, AdminInterface $admin)
+        {
+            return $this->hashId->decodeHex($request->get($admin->getIdParameter()));
+        }
+    }
+
+Add this into your relevant services file and then you will can either
+override this on a per Controller basis where you need it. e.g.
+
+.. code:: php
+
+    class MyCRUDController extends CRUDController
+    {
+        /**
+         * @return int|string
+         */
+        protected function getObjectIdFromRouteIdHandler()
+        {
+            return $this->get('hashid_route_id_handler')
+                ->getIdFromRequest($this->getRequest(), $this->admin);
+        }
+    }
```

Comment:
This is not great. I'd like to have an configuration option to just change the route_id_handler globally (or better on an admin-per-admin basis). If you choose the former, then you will have to check the `Admin` instance to see if decoding is needed or not but if yes, so be it.